0010-zkvm-keccak.md

```diff
@@ -0,0 +1,778 @@
+| -------- | ----------- | ----------- | ------------ | ------------- | ------------- | ------------- | ----- | ---- | 
+| `Curr` | 100`Reset` | 300`Sparse` | (100`Reset`) | 100`Bytes16` | 100`Bytes16` | 100`Bytes16` | 100`Reset` | (100`Reset`) |
+| PiRho    | reset0_e | reseti_e   | dense_e     | quotient_e   | remainder_e   | bound_e       | dense_rot_e   | expand_rot_e |
+
+| Columns: | [1540...1940) | [1940...2340) | 
+| -------- | ------------- | ------------- |
+| `Curr`   | 400`Sparse`   | 400`Sparse`   |
+| Chi      | reset_b       | reset_sum     |
+
+The design makes use of a smaller table containing all bytes (values up to 8 bits), or reuses `Bytes16` or `Bytes12` with a scaling factor.
+
+<center>
+
+| row |
+| --- |
+| $0$ |
+| ... |
+| $255$ |
+
+</center>
+
+The `KeccakSponge` gate performs 200*2 + 100 + 300 lookups:
+
+| `KeccakSponge` | [200...400) | [400...500)  | [500...800)      |
+| -------------- | ---------------- | ------------- | ------------------ | 
+| Curr           | 200`Bytes` + (100`Reset`) | 100`Reset` | 300`Sparse` |
+
+The 100 lookups to `Reset` are performed using a linear combination of the bytes to obtain the corresponding 16 dense bits for each `shift0` term. Take into account that bytes shall be laid out in big-endian format, whereas each dense pair is in little-endian.
+
+$$shift_0[i] = expand(\ bytes[2\cdot i] + 2^8 \cdot bytes[2\cdot i+1]\ )$$
+
+### Performance
+
+Counting the costs of the steps presented above, the following table summarizes the performance of the proposed design, for each of the 24 rounds of the permutation function, in each design of the Keccak gadget.
+
+<center>
+
+| Version  | Columns | Rows / block | Lookups / block | 
+|----------|---------|--------------|-----------------|
+| This RFC | 2344    | $1+24+1(+1)$ | $24\times(180+800+800)+800=43,520$ |    
+
+</center>
+
+## Test plan and functional requirements
+
+1. Testing goals and objectives: 
+    * The resulting Keccak should behave just like the [implementations](https://keccak.team/software.html) suggested by the Keccak original team.
+    * The performance of the system should not be affected.
+    * It will produce the expected block hash.
+2. Testing approach: 
+    * Expected digests are obtained upon input pre-images.
+3. Testing scope: 
+    * Thorough unit tests of the gadget and its components (XOR, AND, NOT, rotation).
+4. Testing requirements: 
+    * When the project is in a more advanced state, run integration tests with other parts of the system (MIPS ISA Implementation, and sys-call for blocks via Cannon).
+5. Testing resources: 
+    * Run the official [test vectors](https://keccak.team/archives.html).
+    * Obtain actual examples of blocks to hash, with different lengths.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This gate uses much longer lookup tables. Understand if the gains in the number of rows of Keccak compensate for the necessity of a $2^{16}$ lookup table.
```

Comment:
We want a 2^16 lookup table for the other instructions anyway; this is a non-concern.